### PR TITLE
FlyteFile and directory updates

### PR DESCRIPTION
**Pull request by wild-endeavor**
Signed-off-by: wild-endeavor <wild-endeavor@users.noreply.github.com>

# TL;DR
Just cleaning up the behavior around files and directories a bit, and updating the documentation.  The main changes are

* Add a `trigger_download` function to trigger downloading if you don't want to explicitly call `open()`.
* Removing the transformer in `os.PathLike` from the main type engine file since now we'll be giving the user a `FlyteFile` object even when the signature is just `os.PathLike`.
* Moving support for `os.PathLike` into the `FlyteFile` transformer.

Please see the updated docstrings for more information.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin
 - [x] Refactor

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

* Added an option to the type engine so transformers can declare to flytekit that they support multiple python types.

### CODE REVIEW

File: flytekit/types/file/file.py [hunk 1]
```diff
@@ -217,46 +248,78 @@ def to_literal(
...
-                meta = BlobMetadata(type=self._blob_type(format=self.get_format(python_type)))
                 return Literal(scalar=Scalar(blob=Blob(metadata=meta, uri=python_val._remote_source)))
 
-            source_path = python_val.path
-            if python_val.remote_path is False:
-                # If the user specified the remote_path to be False, that means no matter what, do not upload
+            # If the user specified the remote_path to be False, that means no matter what, do not upload. Also if the
+            # path given is already a remote path, say https://www.google.com, the concept of uploading to the Flyte
+            # blob store doesn't make sense.
+            if python_val.remote_path is False or ctx.file_access.is_remote(source_path):
```
Comment (kumare3) [inline flytekit/types/file/file.py:266]: should we do the is_remote check? Why not just upload?

File: flytekit/types/file/file.py [hunk 2]
```diff
@@ -184,6 +209,12 @@ def remote_source(self) -> str:
         """
         return self._remote_source
 
+    def trigger_download(self):
```
Comment (kumare3) [inline flytekit/types/file/file.py:212]: Should we use this in __fspath__ instead of self._downloader()?

```
def __fspath__(self):
        # This is where a delayed downloading of the file will happen
        if not self._downloaded:
            self._downloader()
            self._downloaded = True
        return self._path
```

Also I think we need to add the `download cache` check here too!

File: flytekit/types/file/file.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/types/file/file.py:266]: But if a task
```
def t1() -> FlyteFile:
  return FlyteFile("https://www.google.com")
```
what does it mean to upload that?

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -290,17 +289,19 @@ class TypeEngine(typing.Generic[T]):
...
         This should be used for all types that respond with the right type annotation when you use type(...) function
         """
-        if transformer.python_type in cls._REGISTRY:
-            existing = cls._REGISTRY[transformer.python_type]
-            raise ValueError(
-                f"Transformer {existing.name} for type {transformer.python_type} is already registered."
-                f" Cannot override with {transformer.name}"
-            )
-        cls._REGISTRY[transformer.python_type] = transformer
+        types = [transformer.python_type, *(additional_types or [])]
```
Comment (kumare3) [inline flytekit/core/type_engine.py:296]: not needed,,just register multiple times

File: flytekit/types/directory/types.py [hunk 4]
```diff
@@ -128,6 +175,15 @@ def remote_source(self) -> str:
         """
         return self._remote_source
 
+    def trigger_download(self):
```
Comment (kumare3) [inline flytekit/types/directory/types.py:178]: can we return the path and call the method  `download`

File: flytekit/core/type_engine.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/core/type_engine.py:296]: actually i don't know how to do this cleanly.   Because the transformer specifies the python type when it inherits from the base transformer: `class FlyteFilePathTransformer(TypeTransformer[FlyteFile])`, I'd have to create a base class out of the transformer, and then create two child classes out of it I think.  I think this is actually cleaner.

File: flytekit/core/type_engine.py [hunk 5]
```diff
@@ -290,17 +289,19 @@ class TypeEngine(typing.Generic[T]):
     _DATACLASS_TRANSFORMER: TypeTransformer = DataclassTransformer()
 
     @classmethod
-    def register(cls, transformer: TypeTransformer):
+    def register(cls, transformer: TypeTransformer, additional_types: typing.Optional[typing.List[Type]] = None):
```
Comment (cosmicBboy) [inline flytekit/core/type_engine.py:292]: consider docstring for `additional_types` (and maybe `transformer` too?)

**Review by cosmicBboy — APPROVED** [review]
